Added a new event on volume change

mammo0 opened the pull request:
I've added a new event when the audio source has changed the volume. It's similar to the start and stop events before and after the playback.

After the volume has changed a program (or script) can be executed. The program which should be started can be specified in the configuration file with the option `run_this_on_volume_change` in the `sessioncontrol` section. Another way to specify the executable program is to call shairport-sync with the `-C` or `--on-vol-change` argument.

The only limitation when launching the external program is that no user specific arguments can be passed (like it is possible with the start and stop events). Just the new volume value will be passed to the program. This is a double value between 0 and 1.
E.g.: A sample call of the program or script will look like this: `/path/to/program 0.21452`

I also changed the calculation of the `linear_volume` value in the `player_volume(double f)` method (`player.c`).
In my opinion the value wasn't linear calculated before. It was more exponentially (e.g. the value increased faster the higher the volume gets).
With my change the volume is now really linear calculated (at least what I understand under linear).

mikebrady commented:
Hi there. Thank you for all your work – it must have taken quite some time, and I really appreciate your impulse to improve Shairport Sync. I do have a number of serious reservations about your suggestions though. Let me take them in the order of most important first, least important last:
 1. Your suggestion to change the way the linear volume is calculated is based, I think, on a misunderstanding. The general idea is to take the volume level supplied by the source, which is essentially logarithmic in the range 0 dB to -30 dB, and transform it into an acoustically more pleasing volume-change curve – that's what `vol2attn` ("Volume to Attenuation") does, taking the 0 to -30db input and scaling the output in a piecewise-linear fashion to the real volume control's range in decibels*100. If a hardware mixer with dB volume settings is present, the output of the `vol2attn` is passed to it, otherwise it is exponentiated to a linear volume in the range 0 to 65535 (approximately a 48.1 dB range). That is what is meant to happen. Suffice it to say that the progressiveness of the volume control at present is pretty good, and much better than in the original Shairport.
 2. The ability to wait while the command associated with the volume control change executes is dangerous – it could hold up the entire audio processing thread, which is pretty time-sensitive. This is very undesirable, I would say. Even the delay associated with forking a process and then continuing seems to me to be quite considerable. In the same vein, what happens if you get a rapid sequence of volume control changes? Would each one fork separate processes, and could they all be running at the same time?
 3. I really want to stop using command line options, and have stated that future enhancements will only be accessible via the configuration file, so I wouldn't be in favour of the extra command line option you suggested. 
 4. Would it be possible to get the same functionality from a metadata watcher – one that waited on volume control change messages? It seems to me that it would be cleaner.

I really appreciate your work and suggestions, and I'd be glad to hear your responses.
Regards
Mike

mammo0 commented:
Hey, thanks for your fast reply.
First of all please let me explain why I modified shairport-sync (sorry that I haven't done it before):

Currently I'm running XBian on my Raspberry Pi, which is connected to my home cinema system (Pi -> AVR -> TV). Mainly I use the Pi for watching movies on the TV. But when the TV is off, I want to use the Pi as an Airplay receiver. So this is the point where shairport comes to play.
What I want to do is the following: When I start a playback from an audio source I want that my AVR turns on (no problem with the built in run-on-start-command and a little bash-script). And it should turn off after the audio source closes the connection and there is no new connection within five minutes (also no problem).
So now the interesting point. When the audio source changes the volume level, I want the AVR to change its volume as well. Therefore I needed an event which is triggered when the audio source is changing its volume.
I hope that I could make my point a little bit clearer. If you have any questions, feel free to ask.

Now to your four points. I will answer each in the same sequence:
  1. Em ok... :laughing:
I don't have any theoretical knowledge of the Airplay protocol, volume curves etc. So I don't really understand what you were trying to explain me (maybe my English is also too bad for this).
But I think that you have a lot more knowledge in this area. My intention was a little bit more practical: If the volume of the audio source is relatively low and you change it to the higher next step, the reaction of my AVR was not measurable because the `linear_volume` value changes from e.g. 0.00231 to 0.00315. But if the volume of the audio source is high, it makes a large step e.g. from 0.84641 to 1.00000. Unfortunately this behavior is not practicable with the way I want to use shairport. Therefore I changed the calculating of the volume value that it makes equal steps from 0 to 1.
  2. I understand your point. But I have not experienced any issues with the audio processing thread. Even if I have some `sleep` commands in my script (and if I wait for them), the audio is playing well without any interruptions. And with multiple volume changes within a short period of time the script is called only when the script call before has finished.
How this is done in the background I've got now idea (forks, etc.)... Because I didn't have much experience with C-programming. I copied a big part from your code (`command_start()` method) and the rest was some searching on Google. So I think that part can be done much better. But like I said, I don't have the knowledge to do this. (At least not now :wink:)
  3. It's no problem if you want no command line options. I basically implemented the option because there were also two for the start and end events. So you don't have to adopt it.
  4. I've no idea what the metadata watcher is :laughing:. But if you think that is a better way to implement such a functionality I would appreciate if you don't mind to do it. Thanks in advance!

After all I have to say that it's apparently a little bit more complex than I thought before...
So I think this pull request can be closed. But it would be great if you can handle it as a future request for shairport-sync.

Best regards, mammo0

mikebrady commented:
Thanks for your replies. I understand the usage scenario now.
So:
 1. Here's a much better explanation than I could give: http://www.dr-lex.be/info-stuff/volumecontrols.html. Also, the thing about attenuation curves comes from http://tangentsoft.net/audio/atten.html, though I don't pretend the implementation here is wonderful.

 2. Actually I was mistaken – the volume control script would not run on the playback thread – it plays on the main thread handling incoming commands etc. So, not so time-critical, but important not to block nonetheless.

 3. Yep, that's easy alright.

 4. Shairport Sync can be compiled to handle metadata and can send this metadata over a unix pipe. Among the items of metadata are volume control changes. So, one could imagine a separate little program reading the pipe and responding to volume control changes. There is a sample program that does just that at https://github.com/mikebrady/shairport-sync-metadata-reader.

The advantage of this approach is that it is completely independent of Shairport Sync itself, and likely to be pretty compact. What do you think?

mammo0 commented:
Thank you for the provided information. Now I understand what you where trying to explain me. And I also understand why I had such a strange behaviour with your calculated value. It's because my AVR uses a exponentially 'volume slider' too. Therefore I had twice the effect of exponentially growing audio.
So all I have to do is to reverse the exponential value back to a linear in my volume script.

I also took a look at your metadata reader/watcher. I think this is exactly the thing I've searched for. I could change my script that it watches the pipe for volume changes. I think you're right, it's the best way to do it.

Well, than nothing is left from my pull request... So I close it here.
Thank you for time and I hope you continue your work.

mikebrady commented:
Thanks again for your interest and inputs, and for the kind words.
